drivers/sensors/apds9922.c

```diff
@@ -0,0 +1,2619 @@
+
+#define PS_LED_PULSES_MASK      (0x0fff)
+#define PS_SET_LED_PULSES(p)    ((p) & PS_LED_PULSES_MASK)
+
+/* APDS922_PS_MEAS_RATE 0x03 */
+
+#define PS_RESOLUTION_SHIFT     (3)
+#define PS_RESOLUTION_MASK      (3 << PS_RESOLUTION_SHIFT)
+#define PS_SET_RESOLUTION(r)    (((r) << PS_RESOLUTION_SHIFT) | 0x20)
+#define PS_MEASURERATE_SHIFT    (0)
+#define PS_MEASURERATE_MASK     (7 << PS_MEASURERATE_SHIFT)
+#define PS_SET_MEASURERATE(r)   ((r) << PS_MEASURERATE_SHIFT)
+
+/* APDS922_ALS_MEAS_RATE 0x04 */
+
+#define ALS_RESOLUTION_SHIFT    (4)
+#define ALS_RESOLUTION_MASK     (7 << ALS_RESOLUTION_SHIFT)
+#define ALS_SET_RESOLUTION(r)   ((r) << ALS_RESOLUTION_SHIFT)
+#define ALS_MEASURERATE_SHIFT   (0)
+#define ALS_MEASURERATE_MASK    (7 << ALS_MEASURERATE_SHIFT)
+#define ALS_SET_MEASURERATE(r)  ((r) << ALS_MEASURERATE_SHIFT)
+
+/* APDS922_ALS_GAIN 0x05 */
+
+#define ALS_GAIN_SHIFT          (0)
+#define ALS_GAIN_MASK           (7 << ALS_GAIN_SHIFT)
+#define ALS_SET_GAIN(g)         ((g) << ALS_GAIN_SHIFT)
+
+/* APDS_ALS_MAIN_STATUS 0x07 */
+
+#define ALS_INT_STATUS          (16)
+#define ALS_NEW_DATA            (8)
+#define PS_LOGIC_STATUS         (4)
+#define PS_INT_STATUS           (2)
+#define PS_NEW_DATA             (1)
+
+/* APDS9922_PS_DATA0 0x08 */
+
+#define PS_DATA_OVERFLOW_SHIFT  (3)
+#define PS_DATA_OVERFLOW        (1 << PS_DATA_OVERFLOW_SHIFT)
+
+/* APDS9922_INT_CFG 0x19 */
+
+#define PS_INT_EN_SHIFT         (0)
+#define PS_INT_EN               (1 << PS_INT_EN_SHIFT)
+#define PS_INT_MASK             (1 << PS_INT_EN_SHIFT)
+#define PS_LOGIC_MODE_SHIFT     (1)
+#define PS_LOGIC_MODE_NORMAL    (0 << PS_LOGIC_MODE_SHIFT)
+#define PS_LOGIC_MODE_LOGIC     (1 << PS_LOGIC_MODE_SHIFT)
+
+#define ALS_INT_EN_SHIFT        (2)
+#define ALS_INT_EN              (1 << ALS_INT_EN_SHIFT)
+#define ALS_INT_MASK            (5 << ALS_INT_EN_SHIFT)
+#define ALS_INT_VAR_SHIFT       (3)
+#define ALS_INT_VAR_MODE        (1 << ALS_INT_VAR_SHIFT)
+#define ALS_INT_THRESH_MODE     (0 << ALS_INT_VAR_SHIFT)
+#define ALS_INT_SRC_SHIFT       (4)
+#define ALS_INT_SRC_MASK        (3 << ALS_INT_SRC_SHIFT)
+#define ALS_INT_SET_SRC(s)      ((s) << ALS_INT_SRC_SHIFT)
+
+/* APDS922_INT_PERSIST 0x1a */
+
+#define ALS_PERSISTANCE_SHIFT   (4)
+#define ALS_PERSISTANCE_MASK    (15 << ALS_PERSISTANCE_SHIFT)
+#define ALS_SET_PERSISTANCE(p)  ((p) << ALS_PERSISTANCE_SHIFT)
+#define ALS_PERSISTANCE_MAX     (255)
+#define PS_PERSISTANCE_SHIFT    (0)
+#define PS_PERSISTANCE_MASK     (15 << PS_PERSISTANCE_SHIFT)
+#define PS_SET_PERSISTANCE(p)   ((p) << PS_PERSISTANCE_SHIFT)
+#define PS_PERSISTANCE_MAX      (255)
+
+/* APDS922_ALS_THRESH_VAR 0x27  */
+
+#define ALS_THRESH_VAR_SHIFT    (0)
+#define ALS_THRESH_VAR_MASK     (7 << ALS_THRESH_VAR_SHIFT)
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+static const uint32_t apds9922_ps_res_table[] =
```

Comment:
I still do not understand why we need a table.
I see two alternative solutions:
1. We keep 
```
enum
 {
   PS_RES8,
   PS_RES9,
   PS_RES10,
   PS_RES11,
 };
```
and calculate reposition at runtime by `ps_res = 256 << res_index;`. Operation seems to be trivial and even cheaper than `apds9922_ps_res_table[res_index];`
2. Change enum to
```
enum
 {
   PS_RES8 = 256,
   PS_RES9 = 512,
   PS_RES10 = 1024,
   PS_RES11 = 2048,
 };
```
and do not use neither array access and bit shift.

The second approach can be applied to all other mapping tables.

Reply:
it was just "a way" of doing it but does seem clumsy I agree!

Reply:
Yeah. If eventually this tuple of three represents a single configuration, that completely make sense.

Reply:
And with the PS data, there are only 2 uses - the bit value for the register (0, 1, 2, 3) and the max value (2^numbits).

The enum constrains the acceptable values to easy-to-read 8/9/10/11 bits, with the convenient enumerations of 0/1/2/3 to use in the set macro. So isn't the table actually perfectly viable and appropriate here?

It could be a const array I suppose?


```
static const int ps_maxval[] = 
{
  256,
  512,
  1024,
  2048,
};

```
if that's preferable?

Then:

`threshold = ps_maxval[res_index];
`

Reply:
On reflection, your first suggestion makes sense. Implementing it now :)